server/api.go

```diff
@@ -62,3 +63,89 @@ func (d dexAPI) DeleteClient(ctx context.Context, req *api.DeleteClientReq) (*ap
 	}
 	return &api.DeleteClientResp{}, nil
 }
+
+func (d dexAPI) CreatePassword(ctx context.Context, req *api.CreatePasswordReq) (*api.CreatePasswordResp, error) {
+	if req.Password == nil {
+		return nil, errors.New("no password supplied")
+	}
+
+	p := storage.Password{
+		Email:    req.Password.Email,
+		Hash:     req.Password.Hash,
+		Username: req.Password.Username,
+		UserID:   req.Password.UserId,
+	}
+	if err := d.s.CreatePassword(p); err != nil {
+		log.Printf("api: failed to create password: %v", err)
+		return nil, err
+	}
+
+	return &api.CreatePasswordResp{}, nil
+}
+
+func (d dexAPI) UpdatePassword(ctx context.Context, req *api.UpdatePasswordReq) (*api.UpdatePasswordResp, error) {
+	if req.Email == "" {
+		return nil, errors.New("no email supplied")
+	}
+	if req.NewHash == nil && req.NewUsername == "" {
+		return nil, errors.New("nothing to update")
+	}
+
+	if req.NewHash != nil {
+		actual, err := bcrypt.Cost(req.NewHash)
+		if err != nil {
+			log.Printf("api: bcrypt.Cost, error: %s", err)
+		}
+		if actual < bcrypt.MinCost {
+			log.Printf("api: hash does not meet MinCost requirement: %d", bcrypt.MinCost)
```

Comment:
Same as above. Return an error instead of logging.

Also we don't want `MinCost` we probably want `DefaultCost`. Min is the absolute minimum allowed bcrypt cost, not the minimal secure cost.